bindings/java/src/org/sleuthkit/datamodel/blackboardutils/FileAttachment.java

```diff
@@ -155,6 +147,62 @@ private String normalizePath(String path) {
 		return adjustedPath;
 	}
 
+	/**
+	 * Finds the attachment file, given the name and path, and returns the
+	 * object id of the matched file.
+	 *
+	 * @param caseDb              Case database.
+	 * @param dataSource          Data source the message was found in.
+	 * @param fileName            Name of attachment file.
+	 * @param parentPathSubstring Partial parent path of the attachment file.
+	 *
+	 * @throws TskCoreException If there is an error in finding the attached
+	 *                          file.
+	 * @return Object id of the matching file. -1 if no suitable match is found.
+	 */
+	private long findAttachmentFile(SleuthkitCase caseDb, String fileName, String parentPathSubstring, Content dataSource) throws TskCoreException {
+
+		// Use the following algorithm to find a matching file:
+		// - Search across all data sources, but if multiple allocated files match, 
+		//     then preference is given to the file on the specified data source.
+		// - If there are no allocated files that match, but there is one unallocated 
+		//     file that matches, then it is returned.
+		// - If there are more then 1 equally suitable matches, then don't return any match.
+		String whereClause = String.format("LOWER(name) = LOWER('%s') AND LOWER(parent_path) LIKE LOWER('%%%s%%')", fileName, parentPathSubstring);
+		List<AbstractFile> matchedFiles = caseDb.findAllFilesWhere(whereClause);
+
+		long bestMatchAllocatedFileObjId = -1;
+		int bestMatchAllocatedFileCount = 0;
+		long bestMatchUnallocatedFileObjId = -1;
+		int bestMatchUnallocatedFileCount = 0;
+
+		for (AbstractFile file : matchedFiles) {
+			if (file.isMetaFlagSet(TskData.TSK_FS_META_FLAG_ENUM.ALLOC)) {
+				bestMatchAllocatedFileCount++;
+				if (dataSource == file.getDataSource()) {
+					bestMatchAllocatedFileCount = 1;
+					bestMatchAllocatedFileObjId = file.getId();
```

Comment:
If there are multiple matches with only 1 being from the target data source, should you return that one? It seems like this is true from your comment above. However, if the target data source file is the first match of 5, lets say, then it will not be appropriately returned (count will be 5, not one).